File: hash/rust/rosetta-image-hash/src/dhash.rs
//! dhash / dhash_vertical: grayscale → Lanczos resize → adjacent-pixel diff.

use image::DynamicImage;

use crate::average::rgb_to_gray;
use crate::internal::{img_rgb, lanczos};
use crate::{Hash, ImageHashError};

pub fn dhash(img: &DynamicImage, hash_size: usize) -> Result<Hash, ImageHashError> {
	if hash_size < 2 {
		return Err(ImageHashError::InvalidHashSize(hash_size));
	}
	let rgb = img_rgb::to_rgb(img);
	let gray = rgb_to_gray(&rgb);
	let resized = lanczos::resize(&gray, hash_size + 1, hash_size);

	let mut bits: Vec<Vec<bool>> = Vec::with_capacity(hash_size);
	for y in 0..hash_size {
		let mut row = Vec::with_capacity(hash_size);
		for x in 0..hash_size {
			row.push(resized[y][x + 1] > resized[y][x]);
		}
		bits.push(row);
	}
	Ok(Hash::from_bits_unchecked(bits))
}

pub fn dhash_vertical(img: &DynamicImage, hash_size: usize) -> Result<Hash, ImageHashError> {
	if hash_size < 2 {
		return Err(ImageHashError::InvalidHashSize(hash_size));
	}
	let rgb = img_rgb::to_rgb(img);
	let gray = rgb_to_gray(&rgb);
	// Resize to width=N, height=N+1
	let resized = lanczos::resize(&gray, hash_size, hash_size + 1);

	let mut bits: Vec<Vec<bool>> = Vec::with_capacity(hash_size);
	for y in 0..hash_size {
		let mut row = Vec::with_capacity(hash_size);
		for x in 0..hash_size {
			row.push(resized[y + 1][x] > resized[y][x]);
		}
		bits.push(row);
	}
	Ok(Hash::from_bits_unchecked(bits))
}

File: hash/js/rosetta-image-hash/src/dhash.ts
import { Hash, ImageHashError, type RgbImage } from "./hash.js";
import { rgbToGray } from "./averageHash.js";
import { toRgb, validateRgbImage } from "./internal/imgRgb.js";
import { resize } from "./internal/lanczos.js";

/**
 * dhash: grayscale → Lanczos to (W=N+1, H=N) → row-wise adjacent-column diff (strict >).
 */
export function dhash(img: RgbImage, hashSize: number): Hash {
  validateRgbImage(img);
  if (!Number.isInteger(hashSize)) {
    throw new ImageHashError("InvalidHashSize", `hashSize must be an integer, got ${hashSize}`);
  }
  if (hashSize < 2) {
    throw new ImageHashError("InvalidHashSize", `hashSize must be >= 2, got ${hashSize}`);
  }
  const rgb = toRgb(img);
  const gray = rgbToGray(rgb.data, rgb.width, rgb.height);
  const resized = resize(gray, rgb.width, rgb.height, hashSize + 1, hashSize);
  // resized is [hashSize rows][hashSize+1 cols] flattened row-major.

  const bits: boolean[][] = [];
  for (let y = 0; y < hashSize; y++) {
    const rowOff = y * (hashSize + 1);
    const row: boolean[] = new Array(hashSize);
    for (let x = 0; x < hashSize; x++) {
      row[x] = resized[rowOff + x + 1] > resized[rowOff + x];
    }
    bits.push(row);
  }
  return new Hash(bits);
}
